add "Your first PR" worksheet (#66)
tidyverse/tidy-dev-day: worksheet/index.qmd

---
format: typst
---

```{=typst}
#set page(
  flipped: true, 
  columns: 2, 
  margin: (x: 0.5in, y: 1in),
  numbering: none
)
```

## Writing your first documentation PR

The goal of this worksheet is to guide you through the steps needed to successfully create a documentation PR (pull request). Empty spaces have been placed on this sheet for your notes.

### 0. Get ready

> 1. `install.packages("pak")`

Reviewer: Would it be possible to make these check boxes?

1. `pak::pak("devtools")`

1. .Set up your GitHub personal access token.
    - `usethis.r-lib.org` -> Articles -> Managing Git(Hub) Credentials

1. Call `usethis::git_sitrep()` and check that:
    - Your name and email address appears in "Git config (global)"
    - Your GitHub user name is found under "Github"

```{=typst}
#v(1.5in)
```

### 1. Find and claim an issue

Browse the Post-Its, looking for something of interest. If you've never done a PR before, we recommend that you start with a documentation issue since there are slightly fewer moving parts. We also encouraged you to team up with a (new) friend and tackle something together!

Once you've picked a Post-it, open the issue and read the details in full. At this point, you might discover the issue is outside your wheelhouse; if so, no problem, just return the Post-it to the wall and try again.

```{=typst}
#colbreak()
#v(3.5in)
```

> 

Reviewer: Can you draw an appropriately sized rectangle here?

### 2. Get the source locally

* **Fork** and clone the repository using this syntax: 

>     - `usethis::create_from_github("{org}/{repo}")` 

Reviewer: How hard would it to be to make this fill-in-the-blanks too? I think ideally it would be a longer line with org/repo in pale grey text, but I appreciate that might be more work than it's worth.

Author: greying out text was a little hard, tried empty spacing instead

    - (e.g `usethis::create_from_github("tidyverse/ggplot2")`)

* `pak::local_install_dev_deps()` to make sure you've got the necessary packages.

* If you're working on a documentation issue, run `devtools::document()` to make sure you have everything necessary to update the docs before you make any changes.
  
* If you're working on a bug or feature, run `devtools::check()` to get a baseline.

> * If running `devtools::documents()` or `devtools::check()` fails, something is probably off with the setup of your machine and you should seek help.

Reviewer: I think you can rewrite this to assume they're working on a documentation PR, i.e.:

* Remove the conditional in the first bullet.
* Remove the second bullet.
* Remove `check()` from the third bullet.

```{=typst}
#pagebreak()
```

### 3. Make the change

* Create a new branch for your changes to live with

>   `usethis::pr_init("very-brief-description")`

Reviewer: Fill-in-the-blanks?

* Make your change. 
  
  * If you're writing documentation, use `devtools::document()` to 
    update the docs.

  * If you're writing code, add a test or two. 
    Iterate quickly with `devtools::load_all()` and `devtools::test()`.

> * `devtools::check()` to make sure all is still well.

Reviewer: Again, assume documentation.

> * Commit the change with your git client.

Reviewer: Should we assume RStudio here and give a bit more advice?

* If the feature is user-facing (i.e. it adds a new feature or fixes a bug), add a bullet to NEWS. Be concise, link to the issue, and tag yourself.

  ```
  * `use_circleci()` creates a `.circleci/config.yaml` config file for CircleCI (#703, @jdblischak).

>   ```

Reviewer: Can delete since its documentation.

```{=typst}
#v(1in)
```

### 4. Submit your PR

Push and make a PR with `usethis::pr_push()`.
  
- Make the title descriptive:
  - Not good: "Working on issue 45"

>   - Much better: "Use stricter regexes for package and file name checks"

Reviewer: Fill-in-the-blanks?

---
format: typst
---

```{=typst}
#set page(
  flipped: true, 
  columns: 2, 
  margin: (x: 0.5in, y: 1in),
  numbering: none
)

> #set list(marker: ([#sym.square.stroked], [-]))

Reviewer: Could you make this twice as big?

```

## Writing your first documentation PR

> The goal of this worksheet is to guide you through the steps needed to successfully create a documentation PR (pull request). Empty spaces have been placed on this sheet for your notes.

Reviewer: ```suggestion
The goal of this worksheet is to guide you through the steps needed to successfully create your first PR (pull request). Empty spaces have been placed on this sheet for your notes.
```

### 0. Get ready

- `install.packages("pak")`

- `pak::pak("devtools")`

> - .Set up your GitHub personal access token.

Reviewer: ```suggestion
- Set up your GitHub personal access token.
```

    - `usethis.r-lib.org` -> Articles -> Managing Git(Hub) Credentials

- Call `usethis::git_sitrep()` and check that:

>     - Your name and email address appears in "Git config (global)"

Reviewer: Make these check boxes too?

Author: it wasn't pretty, but it works

>     - Your GitHub user name is found under "Github"

Reviewer 2: Not sure if we want to mention this explicitly? But one of the most common mistakes I see at this stage is people don't use `usethis::create_github_token()` and when they get their PAT, they don't pick the right scopes. That will also generate a "tsk, tsk!" from `git_sitrep()`.

We might also want to recommend checking their "git vaccination status" and dealing with it.

Author: i will refer to you, as you have helped many more people deal with git then I ever will

```{=typst}
#v(1.5in)
```

### 1. Find and claim an issue

> Browse the Post-Its, looking for something of interest. If you've never done a PR before, we recommend that you start with a documentation issue since there are slightly fewer moving parts. We also encouraged you to team up with a (new) friend and tackle something together!

Reviewer: ```suggestion
Browse the Post-its, looking for something of interest. Since you've never done a PR before, we recommend starting with a documentation issue since there are slightly fewer moving parts. We also encourage you to team up with a (new) friend and tackle something together!
```

Once you've picked a Post-it, open the issue and read the details in full. At this point, you might discover the issue is outside your wheelhouse; if so, no problem, just return the Post-it to the wall and try again.

```{=typst}
#colbreak()
```

```{=typst}

> #align(center)[

Reviewer: Nice! Just a note that when I printed this from my mac, it was set to auto-scale the page, which reduced it to 96% for some reason.

Author: i hope you are suggesting that we figure out how to print without autoscale instead of setting the dimensions to 3.125 because that just seems wild 😅

  #box(
    height: 3in,
    width: 3in,
    stroke: black,
    [#align(center + horizon)[Place Post-it here]]
  )
]
```

```{=typst}
#v(1in)
```

### 2. Get the source locally

* **Fork** and clone the repository using this syntax: 

>     - `usethis::create_from_github("           /           ")` 

Reviewer: ```suggestion
    - `usethis::create_from_github("________________/________________")` 
```

And similarly elsewhere

    - (e.g `usethis::create_from_github("tidyverse/ggplot2")`)

* `pak::local_install_dev_deps()` to make sure you've got the necessary packages.

* Run `devtools::document()` to make sure you have everything necessary to update the docs before you make any changes.

> * If running `devtools::documents()` fails or produces changes, something is probably off with the setup of your machine and you should seek help.

Reviewer: ```suggestion
* If running `devtools::document()` fails or produces changes, something is probably off with your setup and you should seek help.
```

```{=typst}
#pagebreak()
```

### 3. Make the change

* Create a new branch for your changes to live with

>   * `usethis::pr_init("                      ")`

Reviewer 2: ```suggestion
  * `usethis::pr_init("______________________")`
```

I think this is another good place for a concrete example, like you did with `create_from_github()`

* Make your change. 
  
* use `devtools::document()` to update the docs.

* Commit the change.
  * Check changed files in the git pane.
  * Press `Commit` button.

>   * Write a short message (<80 characters) detailing the change.

Reviewer: Advice about `Fixes #123` should go here

  * Press `Commit` button.

```{=typst}
#v(1in)
```

### 4. Submit your PR

Push and make a PR with `usethis::pr_push()`.
  
- Make the title descriptive:
  - Not good: "Working on issue 45"

>   - Your title: `"                              "`

Reviewer: This should go in the commit message too.

Reviewer 2: Should this have the underscores + a concrete example, as with `create_from_github()` and `pr_init()`?

---
format: typst
---

```{=typst}
#set page(
  flipped: true, 
  columns: 2, 
  margin: (x: 0.5in, y: 1in),
  numbering: none
)

#set list(marker: ([#scale(x: 200%, y: 200%, origin: bottom)[#sym.square.stroked]], [-]))
```

## Writing your first documentation PR

The goal of this worksheet is to guide you through the steps needed to successfully create your first PR (pull request). Empty spaces have been placed on this sheet for your notes.

### 0. Get ready

- `install.packages("pak")`

- `pak::pak("devtools")`

- Set up your GitHub personal access token.
    - `usethis.r-lib.org` -> Articles -> Managing Git(Hub) Credentials

```{=typst}
#set list(marker: [#scale(x: 200%, y: 200%, origin: bottom)[#sym.square.stroked]])
```

- Call `usethis::git_sitrep()` and check that:
    - Your name and email address appears in "Git config (global)"
    - Your GitHub user name is found under "Github"

> 

Reviewer 2: ```suggestion
    - Your GitHub personal access token is discovered and there is no warning about its scopes
    - If `Vaccinated: FALSE`, consider running `usethis::git_vaccinate()`.
```

Once you've picked a Post-it, open the issue and read the details in full. At this point, you might discover the issue is outside your wheelhouse; if so, no problem, just return the Post-it to the wall and try again.

```{=typst}
#colbreak()
```

```{=typst}
#align(center)[
  #box(
    height: 3in,
    width: 3in,
    stroke: black,
    [#align(center + horizon)[Place Post-it here]]
  )
]
```

```{=typst}
#v(1in)
```

### 2. Get the source locally

* **Fork** and clone the repository using this syntax: 
    - `usethis::create_from_github("______________/_______________")` 
    - (e.g `usethis::create_from_github("tidyverse/ggplot2")`)

* `pak::local_install_dev_deps()` to make sure you've got the necessary packages.

* Run `devtools::document()` to make sure you have everything necessary to update the docs before you make any changes.

* If running `devtools::document()` fails or produces changes, something is probably off with your setup and you should seek help.

```{=typst}
#pagebreak()
```

### 3. Make the change

* Create a new branch for your changes to live with
  * `usethis::pr_init("______________________")`

> 

Reviewer 2: ```suggestion
    - (e.g `usethis::pr_init("add-snapshot-tests")`)
```

* Make your change. 
  
* use `devtools::document()` to update the docs.

```{=typst}
#set list(marker: [#scale(x: 200%, y: 200%, origin: bottom)[#sym.square.stroked]])
```

* Commit the change.
  * Check changed files in the git pane.
  * Press `Commit` button.
  * Write a short message (<80 characters) detailing the change. use the [magic fixes keyword](https://help.github.com/en/articles/closing-issues-using-keywords), e.g. "Fixes #123", to link & close your issue.
  * Press `Commit` button.

```{=typst}
#set list(marker: ([#scale(x: 200%, y: 200%, origin: bottom)[#sym.square.stroked]], [-]))
```

```{=typst}
#v(1in)
```

### 4. Submit your PR

Push and make a PR with `usethis::pr_push()`.
  
- Make the title descriptive:
  - Not good: "Working on issue 45"
  - Good: "Use stricter regexes for package and file name checks"

>   - Your title: `"_______________"`

Reviewer 2: ```suggestion
```

- Example with descriptive title and description containing magic keywords: https://github.com/r-lib/usethis/pull/742 

- Write your PR number on the Post-it and move it to the review wall. 

- Ring the gong to celebrate your successful submission!

```{=typst}
#v(0.5in)
```

### 5. Wait for review

> A person from the tidyverse team will try and review your issue as quickly as possible, reading through your PR and offering suggestions for how to improve it. If you want to work on something else while you wait, use `pr_pause()` to pause your pull request while you work on other stuff. It's possible the reviewer might make changes directly to your review - if that happens you can use `pr_pull()` to get their code back onto your computer.

Reviewer 2: I think you need to mention `pr_resume()` after `pr_pause()`, but before `pr_pull()`. It's possible this is more readable as a mini bullet list?

* `pr_pause()` blah blah
* `pr_resume()` blah blah
* `pr_pull()` blah blah
* `pr_finish()` blah blah

They'd only need `pr_pause()` / `pr_resume()` if they plan to work on a different project in the same repo. But I know space is tight here.
